crates/ty_python_semantic/src/program.rs

```diff
@@ -51,41 +56,76 @@ impl Program {
         self.python_version_with_source(db).version
     }
 
+    fn resolve_python_version(
+        config_value: Option<PythonVersionWithSource>,
+        environment_value: Option<&PythonVersionWithSource>,
+    ) -> PythonVersionWithSource {
+        config_value
+            .or_else(|| environment_value.cloned())
+            .unwrap_or_default()
+    }
+
     pub fn update_from_settings(
         self,
         db: &mut dyn Db,
         settings: ProgramSettings,
     ) -> anyhow::Result<()> {
         let ProgramSettings {
-            python_version,
+            python_version: python_version_with_source,
             python_platform,
             search_paths,
         } = settings;
 
+        let search_paths = SearchPaths::from_settings(db, &search_paths)?;
```

Comment:
Could we call `Self::update_search_paths` here instead of duplicating the logic?